drivers/serial/uart_ipc.c

```diff
@@ -0,0 +1,211 @@
+
+#define BIND_TIMEOUT CONFIG_IPC_UART_BIND_TIMEOUT_MS
+
+LOG_MODULE_REGISTER(ipc_uart, CONFIG_IPC_UART_LOG_LEVEL);
+
+struct uart_ipc_config {
+	const struct device *ipc;
+	const char *ept_name;
+};
+
+struct uart_ipc_data {
+	struct ipc_ept ept;
+	struct ipc_ept_cfg ept_cfg;
+	struct k_event ept_bond;
+	struct k_work ipc_send_work;
+	struct ring_buf *tx_ringbuf;
+	struct ring_buf *rx_ringbuf;
+};
+
+static const struct uart_ipc_config *get_config(const struct device *dev)
+{
+	return dev->config;
+}
+
+static struct uart_ipc_data *get_data(const struct device *dev)
+{
+	return dev->data;
+};
+
+static void ipc_send_work_handler(struct k_work *work)
+{
+	int ret;
+	uint32_t data_size;
+	uint8_t *data;
+	struct uart_ipc_data *dev_data;
+
+	dev_data = CONTAINER_OF(work, struct uart_ipc_data, ipc_send_work);
+
+	data_size = ring_buf_get_claim(dev_data->tx_ringbuf, &data,
+				       ring_buf_size_get(dev_data->tx_ringbuf));
+	if (data_size == 0) {
+		/* Nothing to send. */
+		return;
+	}
+
+	ret = ipc_service_send(&dev_data->ept, data, data_size);
+	if (ret == -ENOMEM) {
+		k_work_submit(&dev_data->ipc_send_work);
+	} else if (ret < 0) {
+		__ASSERT(false, "Failed to send data over ipc, err: %d", err);
+		ret = 0;
+	} else {
+		/* Do nothing */
+	}
+
+	ret = ring_buf_get_finish(dev_data->tx_ringbuf, ret);
+	if (ret) {
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+	}
+}
+
+static void ipc_ept_bound(void *priv)
+{
+	const struct device *dev = priv;
+	struct uart_ipc_data *data = get_data(dev);
+
+	k_event_set(&data->ept_bond, 0x01);
+}
+
+static void ipc_ept_data_received(const void *data, size_t len, void *priv)
+{
+	int ret;
+	const struct device *dev = priv;
+	struct uart_ipc_data *dev_data = get_data(dev);
+
+	ret = ring_buf_put(dev_data->rx_ringbuf, data, len);
+	if (ret != len) {
+		LOG_INF("RX ring buffer full. Dropping %d bytes", len - ret);
+	}
+}
+
+static void ipc_ept_error(const char *message, void *priv)
+{
+	LOG_ERR("UART IPC error: %s", message);
+	__ASSERT_NO_MSG(false);
+}
+
+int ipc_uart_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	int ret;
+	struct uart_ipc_data *data = get_data(dev);
+
+	ret = ring_buf_get(data->rx_ringbuf, (uint8_t *)p_char, sizeof(*p_char));
+	if (ret == sizeof(*p_char)) {
+		LOG_DBG("Received character through IPC: %c", *p_char);
+	} else if (ret == 0) {
+		LOG_DBG("Receive buffer is empty");
+	} else {
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+	}
+
+	return (ret == 0) ? (-1) : 0;
+}
+
+void ipc_uart_poll_out(const struct device *dev, unsigned char out_char)
+{
+	int ret;
+	struct uart_ipc_data *data = get_data(dev);
+
+	do {
+		ret = ring_buf_put(data->tx_ringbuf, (uint8_t *)(&out_char), sizeof(out_char));
+		/* We do not want to busy wait here in case when buffer is full. */
+		k_msleep(1);
+	} while (ret < sizeof(out_char));
+
+	/* Should not happen. */
+	if (ret > sizeof(out_char)) {
+		LOG_ERR("Invalid bytes count returned from buffer");
+		__ASSERT_NO_MSG(false);
```

Comment:
Consider using standard assert (with msg) and print an error message in its macro instead of having an individual log.